File: clients/python/baml_lib/_impl/deserializer/raw_wrapper/dict_wrapper.py
import json
import typing
from .raw_wrapper import RawWrapper


T = typing.TypeVar("T", bound=typing.Any)


def filter_empty(
    x: typing.Iterable[typing.Tuple[RawWrapper, typing.Optional[T]]],
) -> typing.Iterable[typing.Tuple[RawWrapper, T]]:
    return typing.cast(
        typing.Iterable[typing.Tuple[RawWrapper, T]],
        filter(lambda v: v[1] is not None, x),
    )


@typing.final
class DictRawWrapper(RawWrapper):
    def __init__(self, val: typing.Mapping[RawWrapper, RawWrapper]) -> None:
        self.__val = val

    def as_self(self) -> typing.Any:
        return {key.as_self(): value.as_self() for key, value in self.__val.items()}

    def as_str(self, inner: bool) -> typing.Optional[str]:
        return json.dumps(self.as_self(), indent=2)

    def as_smart_str(self, inner: bool) -> typing.Optional[str]:
        return self.as_str(True)

    def as_int(self) -> typing.Optional[int]:
        if len(self.__val) == 1:
            for _, item in filter_empty(
                map(lambda kv: (kv[0], kv[1].as_int()), self.__val.items())
            ):
                return item
        return None

    def as_float(self) -> typing.Optional[float]:
        if len(self.__val) == 1:
            for _, item in filter_empty(
                map(lambda kv: (kv[0], kv[1].as_float()), self.__val.items())
            ):
                return item
        return None

    def as_bool(self) -> typing.Optional[bool]:
        if len(self.__val) == 1:
            for _, item in filter_empty(
                map(lambda kv: (kv[0], kv[1].as_bool()), self.__val.items())
            ):
                return item
        return None

    def as_list(self) -> typing.Iterable[RawWrapper]:
        return self.__val.values()

    def as_dict(
        self,
    ) -> typing.ItemsView[typing.Optional[RawWrapper], RawWrapper]:
        return self.__val.items()

File: clients/ts/src/deserializer/raw_wrapper/dict_wrapper.ts
// Import the RawWrapper class
import { RawWrapper } from './raw_wrapper';

// TypeScript doesn't have a direct equivalent of Python's filter_empty function,
// so we'll create a utility function to filter out null or undefined values from an array of tuples
function filterEmpty<T>(array: Array<[RawWrapper, T | undefined | undefined]>): Array<[RawWrapper, T]> {
    return array.filter((item): item is [RawWrapper, T] => item[1] !== null && item[1] !== undefined);
}

// TypeScript equivalent of the DictRawWrapper class
class DictRawWrapper extends RawWrapper {
    private __val: Map<RawWrapper, RawWrapper>;

    constructor(val: Map<RawWrapper, RawWrapper>) {
        super();
        this.__val = val;
    }

    as_self(): any {
        let result: {[key: string]: any} = {};
        this.__val.forEach((value, key) => {
            result[key.as_self()] = value.as_self();
        });
        return result;
    }

    as_str(inner: boolean): string | undefined {
        return JSON.stringify(this.as_self(), null, 2);
    }

    as_smart_str(inner: boolean): string | undefined {
        return this.as_str(true);
    }

    as_int(): number | undefined {
        if (this.__val.size === 1) {
            for (const [key, value] of filterEmpty(Array.from(this.__val.entries()).map(kv => [kv[0], kv[1].as_int()]))) {
                return value;
            }
        }
        return undefined;
    }

    as_float(): number | undefined {
        if (this.__val.size === 1) {
            for (const [key, value] of filterEmpty(Array.from(this.__val.entries()).map(kv => [kv[0], kv[1].as_float()]))) {
                return value;
            }
        }
        return undefined;
    }

    as_bool(): boolean | undefined {
        if (this.__val.size === 1) {
            for (const [key, value] of filterEmpty(Array.from(this.__val.entries()).map(kv => [kv[0], kv[1].as_bool()]))) {
                return value;
            }
        }
        return undefined;
    }

    as_list(): Iterable<RawWrapper> {
        return Array.from(this.__val.values());
    }

    as_dict(): Iterable<[RawWrapper | undefined, RawWrapper]> {
        return this.__val.entries();
    }
}

export { DictRawWrapper };